File: src/helpers/dataValidatorTools.ts
//
// Copyright contributors to the agentic-ai-cyberres project
//
import { DynamicTool, StringToolOutput } from "beeai-framework/tools/base";
import { z } from "zod";
import { ObjectId } from "mongodb";
import * as mongoDB from "mongodb";
import { execSync } from 'child_process';
import { getEnv, parseEnv } from "bee-agent-framework/internals/env";

/*
 * Tool to look at running processs to determine what applications may be running
 */
export const FindRunningProcessesTool = new DynamicTool({
  name: "FindRunningProcesses",
  description: "Determine what applications are running on the system by looking at running processesu. Disregard processes that are used by typical Linux system processes",
  inputSchema: z.object({
    min: z.number().int().min(0),
  }),
  async handler(input) {

	var returnString = new String;
	var stdout = new String;

	// do shell escape to run ps to see what processes are running.  Exclude kernel processes.
	try {
		stdout = execSync('ps  --ppid 2 -p 2 --deselect').toString();
		console.log(`stdout: ${stdout}`);
		returnString = stdout;
	} catch (error: any) {
		console.error(`Error: ${error.message}`);
		if (error.stderr) {
			console.error(`stderr: ${error.stderr.toString()}`);
  		}
		returnString = "Validation Failed.  Details:\n" + error.stderr;
	}

    	return new StringToolOutput(returnString);

  },
});

/*
 * Tool to validate mongoDB databases
 */
export const MongoDBDataValidatorTool = new DynamicTool({
  name: "MongoDBDataValidator",
  description: "This tool validates a mongodb database to ensure the database is not corrupted. Do not use this tool to validate anything that is nto mongoDB. It can only be used if mongoDB is currently running.",
  inputSchema: z.object({
    argument: z.string()
  }),
  async handler(input) {

	var returnString = new String;
	var stdout = new String;
	let workloadToValidate: string = "mongodb";

	console.log(`argument: `+ input.argument);
	// if argument is not mongond, then agent is trying to use this tool to validate another workload.  Don't let it.
	if (input.argument !== workloadToValidate) {
		console.log(input.argument + ` is not a mongoDB workload. This tool cannot be used to validate ` + input.argument);
		returnString = "Validation Failed. Reason: " + input.argument + " cannot be used to validate " + workloadToValidate;
	} else {
		console.log(`Validating mongoDB workload: `+ input.argument);

		// do shell escape to run mongodb commands through mongosh
		try {
			stdout = execSync('mongosh --file src/helpers/mongoDBDataValidator.js').toString();
			console.log(`stdout: ${stdout}`);
			returnString = stdout;
		} catch (error: any) {
			console.error(`Error: ${error.message}`);
			if (error.status) {
				console.error(`Validation failed.`);
			}
			if (error.stderr) {
				console.error(`stderr: ${error.stderr.toString()}`);
  			}
			returnString = "Validation Failed.  Details:\n" + error.stderr;
		}
	}
    	return new StringToolOutput(returnString);
  },
});

/*
 * Tool to send email.
 * Initially intended for emailing results when agent runs autonomously
 */
export const SendEmailTool = new DynamicTool({
  name: "SendEmail",
  description: "Send an email.",
  inputSchema: z.object({
    argument: z.string()
  }),
  async handler(input) {

	var returnString = new String;
	var stdout = new String;
	var email = getEnv("USER_EMAIL");

	// do shell escape to run sendmail to send an email. (for sharing results when agent is run autonomously.))
	try {
		stdout = execSync('set -x; sendmail -t ' + email + ' << EOM\nSubject:Data Validation\n' + input.argument + '\nEOM').toString();
		console.log(`stdout: ${stdout}`);
		returnString = stdout;
	} catch (error: any) {
		console.error(`Error: ${error.message}`);
		if (error.stderr) {
			console.error(`stderr: ${error.stderr.toString()}`);
  		}
		returnString = "Validation Failed.  Details:\n" + error.stderr;
	}

    	return new StringToolOutput(returnString);

  },
});

/*
 * Tool to look at listening ports to determine what applications may be running
 */
export const FindWhatsRunningByPortsTool = new DynamicTool({
  name: "FindWhatsRunningByPorts",
  description: "Determine what applications are running on the system by looking at open listening ports.  Disregard ports that are used by typical Linux system processes.",
  inputSchema: z.object({
    min: z.number().int().min(0),
    max: z.number().int(),
  }),
  async handler(input) {

	var returnString = new String;
	var stdout = new String;

	// do shell escape to run netstat to see what ports are listening
	try {
		stdout = execSync('netstat -al').toString();
		console.log(`stdout: ${stdout}`);
		returnString = stdout;
	} catch (error: any) {
		console.error(`Error: ${error.message}`);
		if (error.stderr) {
			console.error(`stderr: ${error.stderr.toString()}`);
  		}
		returnString = "Validation Failed.  Details:\n" + error.stderr;
	}

    	return new StringToolOutput(returnString);

  },
});

File: python/helpers/data_validator_tools.py
import subprocess
import os
import logging

logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

class StringToolOutput:
    def __init__(self, value: str):
        if not isinstance(value, str):
            raise ValueError("StringToolOutput value must be a string")
        self.value = value

class FindRunningProcessesTool:
    name = "FindRunningProcesses"
    description = "Determine what applications are running on the system by looking at running processes. Disregard processes that are used by typical Linux system processes."

    def __init__(self):
        pass

    def handler(self, input):
        if not isinstance(input, dict):
            logger.error("Input must be a dictionary")
            return StringToolOutput("Validation Failed. Input must be a dictionary.")
        try:
            # Validate input keys if any expected (none expected here)
            stdout = subprocess.check_output(['ps', '-ax', '-o', 'pid,ppid,command'], text=True)
            logger.info("FindRunningProcessesTool executed successfully")
            return StringToolOutput(stdout)
        except subprocess.CalledProcessError as e:
            logger.error(f"Command execution failed: {e}")
            return StringToolOutput(f"Validation Failed. Details:\n{e.output}")
        except Exception as e:
            logger.error(f"Unexpected error: {e}")
            return StringToolOutput(f"Validation Failed. Unexpected error: {str(e)}")

class MongoDBDataValidatorTool:
    name = "MongoDBDataValidator"
    description = "This tool validates a mongodb database to ensure the database is not corrupted. Do not use this tool to validate anything that is not mongoDB. It can only be used if mongoDB is currently running."

    def __init__(self):
        pass

    def handler(self, input):
        if not isinstance(input, dict):
            logger.error("Input must be a dictionary")
            return StringToolOutput("Validation Failed. Input must be a dictionary.")
        argument = input.get("argument", "")
        if not isinstance(argument, str):
            logger.error("Argument must be a string")
            return StringToolOutput("Validation Failed. Argument must be a string.")
        if argument != "mongodb":
            logger.warning(f"Invalid argument for MongoDBDataValidatorTool: {argument}")
            return StringToolOutput(f"Validation Failed. Reason: {argument} cannot be used to validate mongodb")
        try:
            stdout = subprocess.check_output(['mongosh', '--file', 'src/helpers/mongoDBDataValidator.js'], text=True)
            logger.info("MongoDBDataValidatorTool executed successfully")
            return StringToolOutput(stdout)
        except FileNotFoundError:
            logger.error("mongosh command not found")
            return StringToolOutput("Validation Failed. mongosh command not found. Please install mongosh and ensure it is in your PATH.")
        except subprocess.CalledProcessError as e:
            logger.error(f"Command execution failed: {e}")
            return StringToolOutput(f"Validation Failed. Details:\n{e.output}")
        except Exception as e:
            logger.error(f"Unexpected error: {e}")
            return StringToolOutput(f"Validation Failed. Unexpected error: {str(e)}")

class SendEmailTool:
    name = "SendEmail"
    description = "Send an email."

    def __init__(self):
        self.email = os.getenv("USER_EMAIL")
        if not self.email:
            logger.warning("USER_EMAIL environment variable is not set")

    def handler(self, input):
        if not isinstance(input, dict):
            logger.error("Input must be a dictionary")
            return StringToolOutput("Validation Failed. Input must be a dictionary.")
        argument = input.get("argument", "")
        if not isinstance(argument, str):
            logger.error("Argument must be a string")
            return StringToolOutput("Validation Failed. Argument must be a string.")
        if not self.email:
            logger.error("USER_EMAIL environment variable not set")
            return StringToolOutput("Validation Failed. USER_EMAIL environment variable not set.")
        try:
            sendmail_input = f"Subject:Data Validation\n{argument}\n"
            process = subprocess.Popen(['sendmail', '-t', self.email], stdin=subprocess.PIPE, text=True)
            process.communicate(sendmail_input)
            if process.returncode != 0:
                logger.error(f"sendmail command failed with return code {process.returncode}")
                return StringToolOutput(f"Validation Failed. sendmail command failed with return code {process.returncode}")
            logger.info("SendEmailTool executed successfully")
            return StringToolOutput("Email sent successfully.")
        except FileNotFoundError:
            logger.error("sendmail command not found")
            return StringToolOutput("Validation Failed. sendmail command not found. Please install sendmail and ensure it is in your PATH.")
        except Exception as e:
            logger.error(f"Unexpected error: {e}")
            return StringToolOutput(f"Validation Failed. Unexpected error: {str(e)}")

class FindWhatsRunningByPortsTool:
    name = "FindWhatsRunningByPorts"
    description = "Determine what applications are running on the system by looking at open listening ports. Disregard ports that are used by typical Linux system processes."

    def __init__(self):
        pass

    def handler(self, input):
        if not isinstance(input, dict):
            logger.error("Input must be a dictionary")
            return StringToolOutput("Validation Failed. Input must be a dictionary.")
        try:
            stdout = subprocess.check_output(['netstat', '-al'], text=True)
            logger.info("FindWhatsRunningByPortsTool executed successfully")
            return StringToolOutput(stdout)
        except FileNotFoundError:
            logger.error("netstat command not found")
            return StringToolOutput("Validation Failed. netstat command not found. Please install netstat and ensure it is in your PATH.")
        except subprocess.CalledProcessError as e:
            logger.error(f"Command execution failed: {e}")
            return StringToolOutput(f"Validation Failed. Details:\n{e.output}")
        except Exception as e:
            logger.error(f"Unexpected error: {e}")
            return StringToolOutput(f"Validation Failed. Unexpected error: {str(e)}")
